aggregator/elem.go

```diff
@@ -306,20 +325,56 @@ func (e *CounterElem) indexOfWithLock(alignedStart int64) (int, bool) {
 }
 
 func (e *CounterElem) processValue(timeNanos int64, agg aggregation.Counter, fn aggMetricFn) {
-	fn(e.opts.FullCounterPrefix(), e.id, nil, timeNanos, float64(agg.Sum()), e.policy)
+	var fullCounterPrefix = e.opts.FullCounterPrefix()
+	if e.aggOpts.UseDefaultAggregation {
+		// No suffix for default aggregations.
+		fn(fullCounterPrefix, e.id, e.suffix(policy.Sum), timeNanos, float64(agg.Sum()), e.sp)
+		return
+	}
+
+	for _, aggType := range e.aggTypes {
+		fn(fullCounterPrefix, e.id, e.suffix(aggType), timeNanos, agg.ValueOf(aggType), e.sp)
+	}
 }
 
-// NewTimerElem creates a new timer element
-func NewTimerElem(id id.RawID, policy policy.Policy, opts Options) *TimerElem {
-	return &TimerElem{
-		elemBase: elemBase{opts: opts, id: id, policy: policy},
+// NB(cw) suffix overrides default suffixes to not add
+// suffix for default Counter metrics for backward compatibility reasons.
+func (e *CounterElem) suffix(aggType policy.AggregationType) []byte {
+	if aggType == policy.Sum {
+		return nil
+	}
+	return e.opts.Suffix(aggType)
+}
+
+// NewTimerElem creates a new timer element.
+func NewTimerElem(id id.RawID, sp policy.StoragePolicy, aggTypes policy.AggregationTypes, opts Options) *TimerElem {
+	e := &TimerElem{
+		elemBase: newElemBase(opts),
 		values:   make([]timedTimer, 0, defaultNumValues), // in most cases values will have two entries
 	}
+	e.ResetSetData(id, sp, aggTypes)
+	return e
+}
+
+// ResetSetData overrides elemBase.ResetData, to include quantile update for timer elements.
+func (e *TimerElem) ResetSetData(id id.RawID, sp policy.StoragePolicy, aggTypes policy.AggregationTypes) {
+	if aggTypes.IsDefault() {
+		aggTypes = e.opts.DefaultTimerAggregationTypes()
```

Comment:
If we end up making the default agg types configurable for counters and gauges, could make `isAggTypesPooled` to `elemBase` and the logic that determines whether the agg type is pooled into the caller.